File: src/utils/Intents.ts
import BitField from './BitField';

export default class Intents extends BitField {
  public static override FLAGS: Record<string, number> = {
    GUILDS: 1 << 0,
    GUILD_MEMBERS: 1 << 1,
    GUILD_BANS: 1 << 2,
    GUILD_EMOJIS_AND_STICKERS: 1 << 3,
    GUILD_INTEGRATIONS: 1 << 4,
    GUILD_WEBHOOKS: 1 << 5,
    GUILD_INVITES: 1 << 6,
    GUILD_VOICE_STATES: 1 << 7,
    GUILD_PRESENCES: 1 << 8,
    GUILD_MESSAGES: 1 << 9,
    GUILD_MESSAGE_REACTIONS: 1 << 10,
    GUILD_MESSAGE_TYPING: 1 << 11,
    DIRECT_MESSAGES: 1 << 12,
    DIRECT_MESSAGE_REACTIONS: 1 << 13,
    DIRECT_MESSAGE_TYPING: 1 << 14,
    MESSAGE_CONTENT: 1 << 15,
    GUILD_SCHEDULED_EVENTS: 1 << 16,
    GUILD_EMBEDDED_ACTIVITIES: 1 << 17,
    PRIVATE_CHANNELS: 1 << 18,
    CALLS: 1 << 19,
    AUTO_MODERATION_CONFIGURATION: 1 << 20,
    AUTO_MODERATION_EXECUTION: 1 << 21,
    USER_RELATIONSHIPS: 1 << 22,
    USER_PRESENCE: 1 << 23,
    GUILD_MESSAGE_POLLS: 1 << 24,
    DIRECT_MESSAGE_POLLS: 1 << 25,
    DIRECT_EMBEDDED_ACTIVITIES: 1 << 26,
    LOBBIES: 1 << 27,
    LOBBY_DELETE: 1 << 28,
    UNKNOWN_29: 1 << 29,
  };

  public static ALL = Object.values(Intents.FLAGS).reduce((all, p) => all | p, 0);
}

File: src/main/kotlin/com/discord/oauth2rpc/utils/Intents.kt
package com.discord.oauth2rpc.utils

class Intents(bits: Int = 0) : BitField(bits) {
    override val flags: Map<String, Int> get() = Companion.FLAGS

    companion object {
        val FLAGS: Map<String, Int> = mapOf(
            "GUILDS" to (1 shl 0),
            "GUILD_MEMBERS" to (1 shl 1),
            "GUILD_BANS" to (1 shl 2),
            "GUILD_EMOJIS_AND_STICKERS" to (1 shl 3),
            "GUILD_INTEGRATIONS" to (1 shl 4),
            "GUILD_WEBHOOKS" to (1 shl 5),
            "GUILD_INVITES" to (1 shl 6),
            "GUILD_VOICE_STATES" to (1 shl 7),
            "GUILD_PRESENCES" to (1 shl 8),
            "GUILD_MESSAGES" to (1 shl 9),
            "GUILD_MESSAGE_REACTIONS" to (1 shl 10),
            "GUILD_MESSAGE_TYPING" to (1 shl 11),
            "DIRECT_MESSAGES" to (1 shl 12),
            "DIRECT_MESSAGE_REACTIONS" to (1 shl 13),
            "DIRECT_MESSAGE_TYPING" to (1 shl 14),
            "MESSAGE_CONTENT" to (1 shl 15),
            "GUILD_SCHEDULED_EVENTS" to (1 shl 16),
            "GUILD_EMBEDDED_ACTIVITIES" to (1 shl 17),
            "PRIVATE_CHANNELS" to (1 shl 18),
            "CALLS" to (1 shl 19),
            "AUTO_MODERATION_CONFIGURATION" to (1 shl 20),
            "AUTO_MODERATION_EXECUTION" to (1 shl 21),
            "USER_RELATIONSHIPS" to (1 shl 22),
            "USER_PRESENCE" to (1 shl 23),
            "GUILD_MESSAGE_POLLS" to (1 shl 24),
            "DIRECT_MESSAGE_POLLS" to (1 shl 25),
            "DIRECT_EMBEDDED_ACTIVITIES" to (1 shl 26),
            "LOBBIES" to (1 shl 27),
            "LOBBY_DELETE" to (1 shl 28),
            "UNKNOWN_29" to (1 shl 29),
        )

        val ALL: Int = FLAGS.values.reduce { a, b -> a or b }
    }
}
